Sources/Sentry/SentryEnvelopeRateLimit.m

```diff
@@ -0,0 +1,67 @@
+#import <Foundation/Foundation.h>
+#import "SentryEnvelopeRateLimit.h"
+#import "SentryRateLimits.h"
+#import "SentryEnvelope.h"
+#import "SentryRateLimitCategoryMapper.h"
+
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface SentryEnvelopeRateLimit ()
+
+@property(nonatomic, strong) id <SentryRateLimits> rateLimits;
+
+@end
+
+@implementation SentryEnvelopeRateLimit
+
+- (instancetype) initWithRateLimits:(id<SentryRateLimits>) sentryRateLimits {
+    if (self = [super init]) {
+        self.rateLimits = sentryRateLimits;
+    }
+    return self;
+}
+
+- (SentryEnvelope *)removeRateLimitedItems:(SentryEnvelope *)envelope {
+    SentryEnvelope *result = envelope;
+    
+    NSArray<SentryEnvelopeItem *> *itemsToDrop = [self getEnvelopeItemsToDrop:envelope.items];
+    
+    if (itemsToDrop.count > 0) {
+        NSArray<SentryEnvelopeItem *> *itemsToSend = [self getItemsToSend:envelope.items withItemsToDrop:itemsToDrop];
+        
+        result = [[SentryEnvelope alloc] initWithId:envelope.header.eventId items:itemsToSend];
+    }
+    
+    return result;
+}
+
+- (NSArray<SentryEnvelopeItem *> *) getEnvelopeItemsToDrop:(NSArray<SentryEnvelopeItem *> *)items  {
+    NSMutableArray<SentryEnvelopeItem *> *itemsToDrop = [NSMutableArray new];
```

Comment:
L: Since most envelopes we capture (right now I believe all) are single item envelopes, and considering most cases we are not going to drop anything due to rate-limit, we could optmize by avoiding an allocation here:

```suggestion
    if (items.count == 0) {
        return nil;
    }
    NSMutableArray<SentryEnvelopeItem *> *itemsToDrop = [NSMutableArray new];
```

We'd need to allow returning `nil` and consider that on the call site

Reply:
You are right, but currently, I am more concerned about the multiple serializations. We should start there.